src/main/scala/org/ergoplatform/modifiers/state/UtxoSnapshotManifest.scala

```diff
@@ -0,0 +1,94 @@
+package org.ergoplatform.modifiers.state
+
+import com.google.common.primitives.Ints
+import org.ergoplatform.modifiers.ErgoPersistentModifier
+import org.ergoplatform.modifiers.history.Header
+import org.ergoplatform.settings.{Algos, Constants}
+import scorex.core.ModifierTypeId
+import scorex.core.serialization.Serializer
+import scorex.core.validation.ModifierValidator
+import scorex.crypto.authds.{ADDigest, ADKey}
+import scorex.crypto.authds.avltree.batch.serialization.{BatchAVLProverManifest, BatchAVLProverSerializer, ProxyInternalNode}
+import scorex.crypto.authds.avltree.batch.{InternalProverNode, ProverNodes}
+import scorex.crypto.hash.Digest32
+import scorex.util.{ModifierId, bytesToId, idToBytes}
+
+import scala.util.Try
+
+case class UtxoSnapshotManifest(proverManifest: BatchAVLProverManifest[Digest32, Algos.HF],
+                                blockId: ModifierId,
+                                emissionBoxIdOpt: Option[ADKey])
+  extends ErgoPersistentModifier with ModifierValidator {
+
+  override type M = UtxoSnapshotManifest
+
+  override val modifierTypeId: ModifierTypeId = UtxoSnapshotManifest.modifierTypeId
+
+  override lazy val serializer: Serializer[UtxoSnapshotManifest] = UtxoSnapshotManifestSerializer
+
+  override lazy val sizeOpt: Option[Int] = Some(bytes.length)
+
+  lazy val rootDigest: ADDigest = {
+    val (root, height) = proverManifest.oldRootAndHeight
+    ADDigest @@ (root.label :+ height.toByte)
+  }
+
+  override def serializedId: Array[Byte] = UtxoSnapshot.rootDigestToSerializedId(rootDigest)
+
+  override def parentId: ModifierId = blockId
+
+  def chunkRoots: Seq[ADDigest] = {
+    def loop(node: ProverNodes[Digest32]): Seq[ADDigest] = node match {
+      case n: ProxyInternalNode[Digest32] =>
+        Seq(n.leftLabel, n.rightLabel).map(ADDigest !@@ _)
+      case n: InternalProverNode[Digest32] =>
+        loop(n.left) ++ loop(n.right)
```

Comment:
It's better to implement `loop` as a @tailrec function to avoid possible stack overflow